delfin/exporter/prometheus/prometheus.py

```diff
@@ -41,26 +51,63 @@
      values={1600998817585: 20.264160223426305})]
 """
 
-unit_of_metric = {'response_time': 'ms', 'throughput': 'IOPS',
-                  'read_throughput': 'IOPS', 'write_throughput': 'IOPS',
-                  'bandwidth': 'MBps', 'read_bandwidth': 'MBps',
-                  'write_bandwidth': 'MBps'
-                  }
-
 
 class PrometheusExporter(object):
 
+    def __init__(self):
+        self.timestamp_offset_ms = 0
+        self.metrics_dir = cfg.CONF.PROMETHEUS_EXPORTER.metrics_dir
+
+    def check_metrics_dir_exists(self, directory):
+        try:
+            if not os.path.exists(directory):
+                os.makedirs(directory)
+            return True
+        except Exception as e:
+            msg = six.text_type(e)
+            LOG.error("Error while creating metrics directory. Reason: %s",
+                      msg)
+            return False
+
+    def set_timestamp_offset_from_utc_ms(self):
+        """Set timestamp offset from utc required for all metrics"""
+        try:
+            timez = get_localzone()
+            if cfg.CONF.PROMETHEUS_EXPORTER.timezone != 'local':
+                timez = pytz.timezone(cfg.CONF.PROMETHEUS_EXPORTER.timezone)
+            timez.utcoffset(datetime.datetime.now())
+            return int(timez.utcoffset(
+                datetime.datetime.now()).total_seconds() * 1000)
+        except Exception:
+            LOG.error('Error while setting timestamp'
+                      ' offset for prometheus exporter')
+            # return no offset in case of an error
+            return 0
+
     # Print metrics in Prometheus format.
-    def _write_to_prometheus_format(self, f, metric, labels, values):
-        f.write("# HELP storage_%s storage metric for %s\n" % (metric, metric))
-        f.write("# TYPE storage_%s gauge\n" % metric)
+    def _write_to_prometheus_format(self, f, metric,
+                                    labels, prom_labels, values):
+        f.write("# HELP %s  metric for resource %s and instance %s\n"
+                % (metric, labels.get('resource_type'),
+                   labels.get('resource_id')))
+        f.write("# TYPE %s gauge\n" % metric)
 
         for timestamp, value in values.items():
-            f.write("storage_%s{%s} %f %d\n" % (metric, labels,
-                                                value, timestamp))
+            timestamp += self.timestamp_offset_ms
+            f.write("%s{%s} %f %d\n" % (metric, prom_labels,
+                                        value, timestamp))
 
     def push_to_prometheus(self, storage_metrics):
-        with open(cfg.CONF.PROMETHEUS_EXPORTER.metrics_cache_file, "a+") as f:
+        self.timestamp_offset_ms = self.set_timestamp_offset_from_utc_ms()
```

Comment:
In error case, set_timestamp_offset_from_utc_ms returns 0, will that be fine from promethues side?if not we can handle this and next statements as errors and return

Reply:
if we return error this collection will be lost.. if we set 0 atleast there is a chance that prometheus wil receive it and accepets/rejects based on its timezone and rejection time window..